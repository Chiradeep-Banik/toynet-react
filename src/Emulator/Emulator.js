import React from 'react';
import './Emulator.css';
import Instructions from './Instructions/Instructions';
import Tabs from './Tabs/Tabs';
import Visuals from './Visuals/Visuals';

const Emulator = ({ panelData }) => {
  return (
    <div className="emulator">
      <Instructions panelData={panelData} />
      <div>
        <Tabs />
        <Visuals />
      </div>
    </div>
  );
};

export default Emulator;
