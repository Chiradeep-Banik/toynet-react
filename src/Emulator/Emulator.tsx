import React from 'react';
import styled from '@emotion/styled';
import { Box, Flex, Grid } from '@chakra-ui/core';

import Tabs from './Tabs/Tabs';
import Visuals from './Visuals';
import Instructions from './Instructions';
import DialogueBox from './DialogueBox';
import ConsoleTab from './Tabs/ConsoleTab';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  overflow: hidden;
`;


const data = {
  'id': 1,
  'submoduleNumber': 1,
  'submoduleName': 'Modifying Topology',
  'objective': 'Add and remove devices and their connections.',
  'tasks': [
    'Add a Host (h3)',
    'Attach h3 to s2',
    'Detach h3 from s2',
    'Remove h2',
  ],
};

const Emulator = () => {
  return (
    <Grid
      margin='auto'
      maxWidth='1920px'
      height='100vh'
      padding={'0.789rem'}
      gap={2}
      gridTemplateColumns={'auto 1fr'}
    >
      <Instructions panelData={data} />
      <Grid
        gridTemplateRows={'1fr 1fr'}
        width='100%'
        gap={2}
      >
        <Visuals />
        <Grid
          gap={2}
          gridTemplateColumns={'2fr 1fr'}
        >
          <ConsoleTab />
          <DialogueBox />
        </Grid>
      </Grid>
    </Grid>
  );
};

export default Emulator;
