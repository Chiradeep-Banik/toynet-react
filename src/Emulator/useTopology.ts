import { useEffect } from 'react';
import { useImmerReducer } from 'use-immer';

import { DeviceInterface, Action, ReducerFn } from '../common/types';
import { parseXMLTopology, ParsedXML } from '../common/topologyParser';
import { SessionId, CommandRequest } from '../common/api/topology/types';
import { useToynetSession, useModifyTopology } from '../common/api/topology';

interface Connection {
  to: string;
  from: string;
}
type ReducerAction = Action<TopologyActions, DeviceInterface | DeviceInterface[] | ParsedXML | Connection>;

export interface TopologyState {
  hosts: DeviceInterface[];
  routers: DeviceInterface[];
  switches: DeviceInterface[];
  dispatch: React.Dispatch<ReducerAction>;
  isLoading: boolean;
  sessionId: SessionId;
}

export enum TopologyActions {
  ADD_SWITCH,
  ADD_ROUTER,
  ADD_HOST,
  SET_SWITCHES,
  SET_HOSTS,
  SET_ROUTERS,
  ADD_CONNECTION,
  SET_TOPOLOGY,
  CLEAR,

  FLUSH_QUEUE,
}

function reducer(state: ParsedXML, action: ReducerAction) {
  switch (action.type) {
    case TopologyActions.ADD_ROUTER:
    case TopologyActions.ADD_SWITCH:
    case TopologyActions.ADD_HOST:
      const addKey = action.type === TopologyActions.ADD_ROUTER ? 'routers' :
        action.type === TopologyActions.ADD_SWITCH ? 'switches' : 'hosts';
      state[addKey].push(action.payload as DeviceInterface);
      return;

    case TopologyActions.SET_HOSTS:
    case TopologyActions.SET_ROUTERS:
    case TopologyActions.SET_SWITCHES:
      const setKey = action.type === TopologyActions.SET_ROUTERS ? 'routers' :
        action.type === TopologyActions.SET_SWITCHES ? 'switches' : 'hosts';
      state[setKey] = action.payload as DeviceInterface[];
      return;

    case TopologyActions.ADD_CONNECTION:
      const { to, from } = action.payload as Connection;
      const devices = [...state.routers, ...state.hosts, ...state.switches];
      for (const device of devices) {
        if (device.name === to)
          device.connections.push(from);
        if (device.name === from)
          device.connections.push(to);
      }
      return;

    case TopologyActions.SET_TOPOLOGY:
      state.routers = (action.payload as ParsedXML).routers;
      state.switches = (action.payload as ParsedXML).switches;
      state.hosts = (action.payload as ParsedXML).hosts;
      return;
  }
};

// This is probably not the best way to do this, but it's the best way right now.
// we should look at a better way to handle mutations after dispatch in the future.
type CommandFn = (cmd: CommandRequest) => any;
function mutationWrapper (id: SessionId, dispatch: ReducerFn<ReducerAction>, mutate: CommandFn) {
  const queue: CommandRequest[] = [];
  return (action: ReducerAction) => {
    switch (action.type) {
      case TopologyActions.ADD_CONNECTION:
        const { to, from } = action.payload as Connection;
        queue.push({ id, command: `add link ${to} ${from}` });
        mutate({ id, command: `add link ${to} ${from}` });
        break;

      case TopologyActions.ADD_ROUTER:
      case TopologyActions.ADD_SWITCH:
      case TopologyActions.ADD_HOST:
        const addKey = action.type === TopologyActions.ADD_ROUTER ? 'router' :
          action.type === TopologyActions.ADD_SWITCH ? 'switch' : 'host';
        const { name } = action.payload as DeviceInterface;
        queue.push({ id, command: `add ${addKey} ${name}` });
        mutate({ id, command: `add ${addKey} ${name}` });
        break;

      case TopologyActions.FLUSH_QUEUE:
        queue.forEach(mutate);
        queue.splice(0, queue.length);
        break;

      default:
        break;
    }

    dispatch(action);
  };
};

const initialState: ParsedXML = {
  switches: [],
  routers: [],
  hosts: [],
};

/**
 * Provides parsed topology state retrieved from the server.
 */
export function useTopology(id: number) {
  const { data, isLoading } = useToynetSession(id);
  const [mutate] = useModifyTopology(data?.sessionId || -1);
  const [state, dispatch] = useImmerReducer(reducer, initialState);

  useEffect(() => {
    if (data && !isLoading) {
      const res = parseXMLTopology(data.topology);
      dispatch({ type: TopologyActions.SET_TOPOLOGY, payload: res });
    }
  }, [dispatch, isLoading, data]);

  return {
    ...state,
    isLoading,
    dispatch: mutationWrapper(data?.sessionId || -1, dispatch, mutate),
    sessionId: data?.sessionId || -1,
  };
}
