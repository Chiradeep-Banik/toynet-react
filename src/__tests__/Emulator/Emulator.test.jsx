import React from 'react';
import { renderWithTheme } from 'src/common/test-utils/renderWithTheme';
import { act, fireEvent, waitFor } from '@testing-library/react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { withEmulatorAndDialogueProvider } from 'src/Emulator/EmulatorProvider';

jest.mock('src/common/api/topology/requests');

import Emulator from 'src/Emulator/Emulator';


describe('The emulator', () => {
  it('should render and match snapshots', () => {
    const { container } = renderWithTheme(<Emulator />)
    expect(container).toMatchSnapshot();
  })
});