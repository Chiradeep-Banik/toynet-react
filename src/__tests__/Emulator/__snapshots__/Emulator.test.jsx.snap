// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The emulator should render and match snapshots 1`] = `
<div>
  <div
    class="css-1af67e0"
  >
    <div
      class="css-74aybn"
    >
      <div
        class="css-1nxzogq"
      >
        <div
          class="css-ebb1fz"
        >
          <a
            class="css-r59rv1"
            href="/"
          >
            <svg
              class="css-112d3fm"
            >
              backIcon.svg
            </svg>
            <p
              class="css-fzcsno"
            >
              Back to site
            </p>
          </a>
        </div>
        <div
          class="css-1f1lwn6"
        >
          <h2
            class="css-nua0st"
          >
            Module 
            1
          </h2>
          <p
            class="css-fzcsno"
          >
            Modifying Topology
          </p>
        </div>
        <div
          class="css-1f1lwn6"
        >
          <h2
            class="css-nua0st"
          >
            Objective
          </h2>
          <p
            class="css-fzcsno"
          >
            Add and remove devices and their connections.
          </p>
        </div>
        <div
          class="css-1f1lwn6"
        >
          <h2
            class="css-nua0st"
          >
            Tasks
          </h2>
          <div
            class="css-16cm47t"
          >
            <li
              class="css-p5me83"
            >
              Add a Host (h3)
            </li>
            <li
              class="css-p5me83"
            >
              Attach h3 to s2
            </li>
            <li
              class="css-p5me83"
            >
              Detach h3 from s2
            </li>
            <li
              class="css-p5me83"
            >
              Remove h2
            </li>
          </div>
        </div>
      </div>
      <div
        style="margin-top: 2vh; z-index: 0;"
      >
        <div
          class="css-5d2y56"
        >
          <div
            class="css-1c5a9no"
          />
          <button
            class="css-x4n12y"
          >
            Clear
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-1xbpgl1"
    >
      <div>
        <div
          class="css-0"
        >
          <div
            aria-orientation="horizontal"
            class="css-e2etbk"
            role="tablist"
          >
            <button
              aria-controls="panel:1-0"
              aria-selected="true"
              class="css-16futwh"
              id="tab:1-0"
              role="tab"
              tabindex="0"
              type="button"
            >
              Configure
            </button>
            <button
              aria-controls="panel:1-1"
              aria-selected="false"
              class="css-16futwh"
              id="tab:1-1"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Console
            </button>
            <button
              aria-controls="panel:1-2"
              aria-selected="false"
              class="css-16futwh"
              id="tab:1-2"
              role="tab"
              tabindex="-1"
              type="button"
            >
              History
            </button>
          </div>
          <div
            class="css-1vsi4k8"
            tabindex="-1"
          >
            <div
              aria-labelledby="tab:1-0"
              class="css-1jj07fx"
              id="panel:1-0"
              role="tabpanel"
              tabindex="-1"
            >
              <div
                class="css-1j6w374"
              >
                <div
                  class="css-lmdmu1"
                >
                  <div>
                    <div
                      class="css-vl1v3j"
                    >
                      <svg
                        class="css-8s9z6t"
                        data-testid="plus-icon"
                      >
                        add.svg
                      </svg>
                      Router
                    </div>
                    <div
                      class="css-zop8l2"
                    >
                      <div
                        class="css-ixj81f"
                      >
                        <div />
                      </div>
                    </div>
                  </div>
                  <div>
                    <div
                      class="css-vl1v3j"
                    >
                      <svg
                        class="css-8s9z6t"
                        data-testid="plus-icon"
                      >
                        add.svg
                      </svg>
                      Switch
                    </div>
                    <div
                      class="css-zop8l2"
                    >
                      <div
                        class="css-ixj81f"
                      >
                        <div />
                      </div>
                    </div>
                  </div>
                  <div>
                    <div
                      class="css-vl1v3j"
                    >
                      <svg
                        class="css-8s9z6t"
                        data-testid="plus-icon"
                      >
                        add.svg
                      </svg>
                      Host
                    </div>
                    <div
                      class="css-zop8l2"
                    >
                      <div
                        class="css-ixj81f"
                      >
                        <div />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="tab:1-1"
              class="css-1jj07fx"
              hidden=""
              id="panel:1-1"
              role="tabpanel"
              tabindex="-1"
            >
              <h2
                class="css-cumyq4"
              >
                Coming soon...
              </h2>
            </div>
            <div
              aria-labelledby="tab:1-2"
              class="css-1jj07fx"
              hidden=""
              id="panel:1-2"
              role="tabpanel"
              tabindex="-1"
            >
              <h2
                class="css-cumyq4"
              >
                Coming soon...
              </h2>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1w23y4q"
      >
        <div
          class="css-ljj55y"
        >
          <div
            class="react-flow"
          >
            <div
              class="react-flow__renderer react-flow__zoompane"
            >
              <div
                class="react-flow__nodes"
                style="transform: translate(0px,0px) scale(1);"
              />
              <svg
                class="react-flow__edges"
                height="500"
                width="500"
              >
                <defs>
                  <marker
                    class="react-flow__arrowhead"
                    id="react-flow__arrowclosed"
                    markerHeight="12.5"
                    markerWidth="12.5"
                    orient="auto"
                    refX="0"
                    refY="0"
                    viewBox="-10 -10 20 20"
                  >
                    <polyline
                      fill="#b1b1b7"
                      points="-5,-4 0,0 -5,4 -5,-4"
                      stroke="#b1b1b7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1"
                    />
                  </marker>
                  <marker
                    class="react-flow__arrowhead"
                    id="react-flow__arrow"
                    markerHeight="12.5"
                    markerWidth="12.5"
                    orient="auto"
                    refX="0"
                    refY="0"
                    viewBox="-10 -10 20 20"
                  >
                    <polyline
                      fill="none"
                      points="-5,-4 0,0 -5,4"
                      stroke="#b1b1b7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </marker>
                </defs>
                <g
                  transform="translate(0,0) scale(1)"
                />
              </svg>
              <div
                class="react-flow__pane"
              />
            </div>
            <div
              class="react-flow__controls css-1if3jds"
            >
              <div
                class="react-flow__controls-button react-flow__controls-zoomin"
              >
                <svg
                  viewBox="0 0 32 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M32 18.133H18.133V32h-4.266V18.133H0v-4.266h13.867V0h4.266v13.867H32z"
                  />
                </svg>
              </div>
              <div
                class="react-flow__controls-button react-flow__controls-zoomout"
              >
                <svg
                  viewBox="0 0 32 5"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h32v4.2H0z"
                  />
                </svg>
              </div>
              <div
                class="react-flow__controls-button react-flow__controls-fitview"
              >
                <svg
                  viewBox="0 0 32 30"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.692 4.63c0-.53.4-.938.939-.938h5.215V0H4.708C2.13 0 0 2.054 0 4.63v5.216h3.692V4.631zM27.354 0h-5.2v3.692h5.17c.53 0 .984.4.984.939v5.215H32V4.631A4.624 4.624 0 0027.354 0zm.954 24.83c0 .532-.4.94-.939.94h-5.215v3.768h5.215c2.577 0 4.631-2.13 4.631-4.707v-5.139h-3.692v5.139zm-23.677.94a.919.919 0 01-.939-.94v-5.138H0v5.139c0 2.577 2.13 4.707 4.708 4.707h5.138V25.77H4.631z"
                  />
                </svg>
              </div>
              <div
                class="react-flow__controls-button react-flow__controls-interactive"
              >
                <svg
                  viewBox="0 0 25 32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.333 10.667H19.81V7.619C19.81 3.429 16.38 0 12.19 0 8 0 4.571 3.429 4.571 7.619v3.048H3.048A3.056 3.056 0 000 13.714v15.238A3.056 3.056 0 003.048 32h18.285a3.056 3.056 0 003.048-3.048V13.714a3.056 3.056 0 00-3.048-3.047zM12.19 24.533a3.056 3.056 0 01-3.047-3.047 3.056 3.056 0 013.047-3.048 3.056 3.056 0 013.048 3.048 3.056 3.056 0 01-3.048 3.047zm4.724-13.866H7.467V7.619c0-2.59 2.133-4.724 4.723-4.724 2.591 0 4.724 2.133 4.724 4.724v3.048z"
                  />
                </svg>
              </div>
            </div>
            <svg
              class="react-flow__background"
              style="width: 100%; height: 100%;"
            >
              <pattern
                height="16"
                id="pattern"
                patternUnits="userSpaceOnUse"
                width="16"
                x="0"
                y="0"
              >
                <circle
                  cx="0.25"
                  cy="0.25"
                  fill="#aaa"
                  r="0.5"
                />
              </pattern>
              <rect
                fill="url(#pattern)"
                height="100%"
                width="100%"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
